src/kernel/serial.zig

```diff
@@ -0,0 +1,39 @@
+const arch = @import("arch.zig").internals;
+
+const PORT_LCR = 3;
+const COM1 = 0x3F8;
+const BAUD_MAX = 115200;
+const CHAR_LEN = 8;
+const SINGLE_STOP_BIT = true;
+const PARITY_BIT = false;
+
+const SerialError = error {
+    InvalidBaud
+};
+
+fn lcr_val(char_len: u8, comptime stop_bit: bool, comptime parity_bit: bool, msb: u8) u8 {
+    return ((char_len) & 0x3) | ((if (stop_bit) 0 else 1) << 2) | (((if (parity_bit) 1 else 0) & 0x7) << 3) | ((msb & 0x1) << 7);
+}
+
+fn baud_divisor(baud: u32) SerialError!u16 {
+    if (baud > BAUD_MAX or baud <= 0) return SerialError.InvalidBaud;
+    return @intCast(u16, BAUD_MAX / baud);
+}
+
+fn transmit_is_empty() bool {
+    return arch.inb(COM1 + 5) & 0x20 > 0;
+}
+
+pub fn init(baud: u32) SerialError!void {
+    const divisor: u16 = try baud_divisor(baud);
+    arch.outb(PORT_LCR + COM1, lcr_val(0, false, false, 1));
+    arch.outb(COM1, @truncate(u8, divisor));
+    arch.outb(COM1 + 1, @truncate(u8, divisor >> 8));
+    arch.outb(COM1 + PORT_LCR, lcr_val(CHAR_LEN, SINGLE_STOP_BIT, PARITY_BIT, 0));
+    arch.outb(COM1 + 1, 0);
+}
+
+pub fn write(char: u8) void {
+    while (transmit_is_empty() == false) {}
```

Comment:
May also be a good idea to add hlt and sli instructions in the while so that you are not spinning as 100%. Could add this:
pub fn function_name(condition: bool) void {
     while (condition) {
        arch.enableInterrupts();
        arch.halt();
        arch.disableInterrupts();
    }
    arch.enableInterrupts();
}
to arch.zig

Reply:
Could I just call `arch.halt()` in the while?

Reply:
i think that should work :)